### PR TITLE
feat(hardware-testing): Rough gravimetric test for OT2

### PR DESCRIPTION
**Pull request by andySigler**
# Overview

Adds a script to run a gravimetric test on an OT2 fixture. This test is functionally the same as what is in production:
- OT2 pipettes and test fixture
- Radwag scale
- OT2 aspirates/dispenses water from vial on scale
- Stores measured weights
- Calculated %CV at multiple volumes

Implements functionality intended to be shared between future tests:
- Scale streaming and recording
- Liquid class pipetting methods
- Liquid level tracking per Well
- Server and frontend to visualize real-time scale data

# Changelog

This PR adds a lot of functionality, with the idea that almost all of this would be reused by all production pipette testing scripts. This functionality will be continuously updated and reworked as production scripts are developed and iterated on.

# Review requests

It's a lot, and is only used by myself for now, so unless you're interested then just putting your eyes on it is enough for now I think.

# Risk assessment

Quite low.

### CODE REVIEW

**Comment by sfoster1** [comment]
Looks like you've got some lint errors in your plotly render hook. We could ignore these, but if we have the tools set up for linting we might as well take advantage. Here are the errors, which should also be visible on the lines they apply to when looking at the pull request files-changed view. You can run these locally with `make lint-js`.

```
/home/runner/work/opentrons/opentrons/hardware-testing/hardware_testing/tools/plot/index.js
Error:    39:5   error  Unexpected var, use let or const instead                       no-var
Error:    40:5   error  Unexpected var, use let or const instead                       no-var
Error:    53:5   error  'example' is not defined                                       no-undef
Error:    74:5   error  Unexpected var, use let or const instead                       no-var
Error:    76:5   error  Unexpected var, use let or const instead                       no-var
Error:    81:5   error  Unexpected var, use let or const instead                       no-var
Error:    82:5   error  Unexpected var, use let or const instead                       no-var
Error:    83:5   error  Unexpected var, use let or const instead                       no-var
Error:    85:10  error  Unexpected var, use let or const instead                       no-var
Error:    90:9   error  Unexpected var, use let or const instead                       no-var
Error:    91:9   error  Unexpected var, use let or const instead                       no-var
Error:    95:9   error  Unexpected var, use let or const instead                       no-var
Error:    95:13  error  It's not necessary to initialize 'stableGrams' to undefined    no-undef-init
Error:   100:9   error  Unexpected var, use let or const instead                       no-var
Error:   100:13  error  It's not necessary to initialize 'unstableGrams' to undefined  no-undef-init
Error:   110:5   error  Unexpected var, use let or const instead                       no-var
Error:   111:5   error  Unexpected var, use let or const instead                       no-var
Error:   111:9   error  It's not necessary to initialize '_interval' to undefined      no-undef-init
Error:   112:5   error  Unexpected var, use let or const instead                       no-var
Error:   118:5   error  Unexpected var, use let or const instead                       no-var
Error:   119:5   error  Unexpected var, use let or const instead                       no-var
Error:   129:9   error  Unexpected var, use let or const instead                       no-var
Error:   135:9   error  Unexpected var, use let or const instead                       no-var
Error:   137:9   error  'Plotly' is not defined                                        no-undef
Error:   141:9   error  Unexpected var, use let or const instead                       no-var
Error:   149:9   error  Unexpected var, use let or const instead                       no-var
Error:   151:13  error  Unexpected var, use let or const instead                       no-var
Error:   152:13  error  Unexpected var, use let or const instead                       no-var
Error:   155:13  error  Unexpected var, use let or const instead                       no-var
Error:   155:17  error  'newDataPipette' is assigned a value but never used            no-unused-vars
Error:   158:13  error  'Plotly' is not defined                                        no-undef
Error:   165:9   error  Unexpected var, use let or const instead                       no-var
Error:   174:9   error  Unexpected var, use let or const instead                       no-var
Error:   181:25  error  Expected '===' and instead saw '=='                    
``

File: hardware-testing/hardware_testing/drivers/asair_sensor.py [hunk 1]
```diff
@@ -17,6 +18,9 @@
 
 log = logging.getLogger(__name__)
 
+USB_VID = 1027
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/drivers/asair_sensor.py]: kind of a nitpick but usb ids are typically rendered in hex so they're always the same length, e.g. 0x0403 for the vid and 0x6007 for the pid; these are more useful for looking up in online resources, for instance.

File: hardware-testing/hardware_testing/drivers/radwag/driver.py [hunk 2]
```diff
@@ -18,6 +18,13 @@
 class RadwagScaleBase(ABC):
     """Base class if Radwag scale driver."""
 
+    @classmethod
+    def vid_pid(cls) -> Tuple[int, int]:
+        """Radwag scale VID:PID."""
+        # TODO: check and handle the vid:pid of other scales
+        # vid:pid for Radwag's "AS 82/220.X2 PLUS" USB2 (usb-b) port
+        return 1155, 41389
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/drivers/radwag/driver.py]: make them hex maybe?

File: hardware-testing/hardware_testing/execute/gravimetric.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    GravimetricRecording,
+    GravimetricRecorder,
+    GravimetricRecorderConfig,
+)
+from hardware_testing.pipette.liquid_class import PipetteLiquidClass
+from hardware_testing.pipette.timestamp import SampleTimestamps
+
+
+LIQUID_CLASS_LOOKUP = {300: liquid.defaults.DEFAULT_LIQUID_CLASS_OT2_P300_SINGLE}
+SCALE_SECONDS_TO_SETTLE = 15
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/execute/gravimetric.py]: it's not really _necessary_ but can be kind of nice to annotate these as [`Final`](https://docs.python.org/3/library/typing.html#typing.Final) which you might have to import from `typing_extensions`

File: hardware-testing/hardware_testing/labware/layout.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+from pathlib import Path
+from typing import Optional
+
+from opentrons import protocol_api
+from opentrons.protocol_api.labware import Labware
+
+from hardware_testing.labware.definitions import load_radwag_vial_definition
+from hardware_testing.opentrons_api.workarounds import is_running_in_app
+
+APP_TIPRACK_CALIBRATION_SLOT = "8"  # where the App puts the tiprack
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/labware/layout.py]: these could also be `Final`

File: hardware-testing/hardware_testing/labware/layout.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    tiprack_multi: Optional[str]
+    trough: Optional[str]
+    plate: Optional[str]
+    vial: Optional[str]
+
+
+class LayoutLabware:
+    """Layout Labware."""
+
+    def __init__(self, slots: LayoutSlots) -> None:
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/labware/layout.py]: Is there a way you could get rid of `build` and `load` and do everything in `__init__`? if you do, you'll be able to make `_ctx` non-optional

File: hardware-testing/hardware_testing/measure/weight/scale.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        """Scale Class."""
+        self._ctx = ctx
+        self._scale = scale
+
+    @classmethod
+    def build(cls, ctx: ProtocolContext) -> "Scale":
+        """Build."""
+        if ctx.is_simulating():
+            _scale = SimRadwagScale()
+        else:
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/measure/weight/scale.py:66]: what error do we get without the type ignore? does  `_scale` need to be a `Union`

File: hardware-testing/hardware_testing/measure/weight/scale.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        self._scale.connect()
+
+    def disconnect(self) -> None:
+        """Scale connect."""
+        self._scale.disconnect()
+
+    def initialize(self, cfg: ScaleConfig = DEFAULT_SCALE_CONFIG) -> None:
+        """Initialize."""
+        # Some Radwag settings cannot be controlled remotely.
+        # Listed below are the things the must be done using the touchscreen:
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/measure/weight/scale.py:90]: should these be in protocol comments?

File: hardware-testing/hardware_testing/opentrons_api/workarounds.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    labware_imp = labware._implementation
+    labware_delta = labware.calibrated_offset - labware_imp.get_geometry().offset
+    labware.set_offset(
+        x=labware_delta.x + x, y=labware_delta.y + y, z=labware_delta.z + z
+    )
+
+
+def force_prepare_for_aspirate(pipette: InstrumentContext) -> None:
+    """Force prepare for aspirate."""
+    # FIXME: remove this and use latest API version once available
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/opentrons_api/workarounds.py:56]: this is now available after your pr got merged I believe

**Review by sfoster1 — CHANGES_REQUESTED** [review]
This is great! I really like the liquid handling stuff. 

I think we should fix the js lint warnings.

I also think that while some of this stuff is going to be... rough... to test and maybe not worth it right now, we could definitely stand to add tests for other core functionality. The exact boundaries of what should and shouldn't be tested are definitely open for debate, but what I'd suggest is
- The `liquid` and possibly `pipette` functionality, not necessarily the data, should have unit tests
- Everything other than that should probably be added to https://github.com/Opentrons/opentrons/blob/edge/.codecov.yml because otherwise you get really annoying annotations in pull requests like this.

Secondly, we should maybe add some instructions to the readme of some of these modules; there's a lot of little setup instructions stuff in code comments that can be in readmes.

Finally, it looks like this is really simulatable. Given that, even aside from unit tests, it might be worth it to have a kind of integration test, where you use a makefile task and a github action to just straight up run some of these protocols in simulation mode. Don't even have to check the outputs, it's really just running to check that it runs and doesn't crash.

When you do that, and this is critical, you should add changes to the `api`, `shared-data`, and `hardware` subdirectories as triggers for the workflow. That means that when the rest of software changes the parts of the system this depends on, the test will get run, and we won't merge anything that breaks this without fixing it.

Overall this looks great!

**Review by sfoster1 — CHANGES_REQUESTED** [review]
It's looking great! Last thing is the tests failing - I think if you put an empty `__init__.py` in `tests/` and `tests/hardware_testing/` you'll be good. You may also need an empty `conftest.py` in `tests/`.

**Comment by andySigler** [comment]
@sfoster1 added READMEs and a workflow for running protocols (just one protocol for now) 🦖

**Review by sfoster1 — APPROVED** [review]
Looks excellent to me. Just one more js format (it includes markdown) and we'll be good I think.

**Pull request merged**